Add ECS relevant info for container insight receiver

JohnWu20 opened the pull request:
**Description:** 
This commit add the ecs info that are used by container insight receiver for ecs instance level metrics. This includes:

Get task info, clusterName and containerInstanceId from Amazon [ECS container agent introspection](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-introspection.html).

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2307

**Testing:** <Describe what testing was performed and which tests were added.>
Unit test

**Documentation:** <Describe the documentation added.>

JohnWu20 committed:
Add ECS relevant info for container insight reciever

JohnWu20 committed:
Add ECS relevant info for container insight reciever

alolita commented:
Hi @JohnWu20 please fix the failing tests listed below. 

    windows-test - Failed
    integration-tests - Failed
    unit-tests - Failed
    lint - Failed

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+type HttpClientProvider interface {
+	Request(path string, ctx context.Context, logger *zap.Logger) ([]byte, error)
```
```suggestion
	Request(ctx context.Context, path string, logger *zap.Logger) ([]byte, error)
```

`ctx` should always be the first argument.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	client                  clientProvider
+}
+
+type HttpClientProvider interface {
```
```suggestion
type Requester interface {
```

I'm not sure `HttpClientProvider` makes sense as an interface name, it doesn't tell me what the interface does and seems like it would be implemented _by_ an `HTTPClient` and not provide one.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	maxHttpResponseLength          = 5 * 1024 * 1024 // 5MB
+)
+
+type HttpClient struct {
```
```suggestion
type HTTPClient struct {
```

`HTTP` is an initialism and should have consistent capitalization.  This should be addressed in all identifiers that include `HTTP` or `http`.  It might also make sense to simply call this struct `Client` as it will be referenced externally as `httpclient.HTTPClient` or `httpclient.Client`.  Avoiding stutter in scoped names is advised.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	Request(path string, ctx context.Context, logger *zap.Logger) ([]byte, error)
+}
+
+type clientProvider interface {
```
```suggestion
type doer interface {
```

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+
+}
+
+func NewHttp(maxRetries int, backoffRetryBaseInMills int, provider clientProvider) HttpClientProvider {
```
Can this be folded into the `New()` method through the use of [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)?  I'd think this might be a better interface:

```go
client := httpclient.New(WithMaxRetries(5), WithClient(&http.Client{TLSConfig: ...}))
```

For the retry base time, use `time.Duration` rather than `int`, that will also allow you to drop the unit statement from the rather long name.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+func (h *HttpClient) backoffSleep(currentRetryCount int) {
+	backoffInMillis := int64(float64(h.backoffRetryBaseInMills) * math.Pow(2, float64(currentRetryCount)))
+	sleepDuration := time.Millisecond * time.Duration(backoffInMillis)
+	if sleepDuration > 60*1000 {
```
Should the maximum `sleepDuration` be configuration exposed to the user?  If they're given control over retries and backoff timing, why not let them specify how long they're willing to wait?

JohnWu20 replied:
All these parameters will not be changed through the config file, so they will not be controlled by customer.

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(endpoint, ctx)
+		if err != nil {
+			log.Printf("W! retry [%d/%d], unable to get http response from %s, error: %v", i, h.maxRetries, endpoint, err)
```
Why does this method take a `logger` parameter and then use the stdlib `log.Printf()`?  I also wonder whether the logger needs to be provided with each invocation of `Request()` or can it be included in the `HTTPClient` struct and provided once with the invocation of `New()`.

JohnWu20 replied:
Taskinfo and instanceInfo will just call `Request(endpoints) ` of the HTTPClient. And the `httpclient` will be initialize once to use for both of them. So I think it's better to put the logger in the `Request()`
And I will change all the` log.Printf()` to use logger.
Thanks!

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	return
+}
+
+func (h *HttpClient) request(endpoint string, ctx context.Context) ([]byte, error) {
```
```suggestion
func (h *HttpClient) request(ctx context.Context, endpoint string) ([]byte, error) {
```

`ctx` must be the first parameter.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+	return body, nil
+}
+
+func (h *HttpClient) clientGet(url string, ctx context.Context) (resp *http.Response, err error) {
```
```suggestion
func (h *HttpClient) clientGet(ctx context.Context, url string) (resp *http.Response, err error) {
```

`ctx` must be the first parameter.

JohnWu20 replied:
Will change it, thanks

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+func (h *HttpClient) Request(endpoint string, ctx context.Context, logger *zap.Logger) (body []byte, err error) {
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(endpoint, ctx)
+		if err != nil {
```
Not all errors should be retried.  This needs to have some logic for determining which errors are safe to retry and which will never succeed and should result in immediate termination.

JohnWu20 replied:
Changed it using new retryablehttp package.

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+		return nil, fmt.Errorf("unable to get response from %s, status code: %d", endpoint, resp.StatusCode)
+	}
+
+	if resp.ContentLength >= maxHttpResponseLength {
```
Why is this check here?  Hasn't the request already been made and the response received.

JohnWu20 replied:
This httpclient will mainly use for ECS Introspection endpoint. And sometimes, when a task have too many containers, this response will be very large that we don't want to deal with this situation.

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,131 @@
+		return nil, fmt.Errorf("unable to read response body from %s, error: %v", endpoint, err)
+	}
+
+	if len(body) == maxHttpResponseLength {
```
Is this necessarily the case?  What about a response that is exactly `maxHTTPResponseLength` bytes?  Does the `io.LimitReader` not provide an error when the limit is reached or some other mechanism?

JohnWu20 replied:
This part of logic just wants to keep same with the situation that we know the length. It means that we don't want to deal with the response when its body is too big.

Aneurysm9 reviewed (changes requested):
This PR is too large to review and needs to be decomposed.  Let's start with the HTTP client that I've already provided feedback on.  Though I would also say that HTTP clients are well-trod ground and wonder whether there is an existing implementation with retry logic that can be re-used here rather than creating a new implementation.

JohnWu20 committed:
Fix the lint issue and split the code

JohnWu20 committed:
Fix the lint issue and split the code

JohnWu20 commented:
> Hi @JohnWu20 please fix the failing tests listed below.
> 
> ```
> windows-test - Failed
> integration-tests - Failed
> unit-tests - Failed
> lint - Failed
> ```

Thanks Alolita, I have fixed the failure, and currently the 
loadtest - Failed which I believe there is no relation-ship with my code change.
@Aneurysm9  I have split the  to two parts, and here is anther [pr](url)  Please help me review it

Aneurysm9 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,128 @@
+	cii.Unlock()
+
+	if cii.clusterName != "" && cii.containerInstanceID != "" {
+		close(cii.readyC)
```
What ensures this isn't closed twice?  Is there any way to eliminate this channel?  It seems to only be used for tests, which is a nasty code smell.

JohnWu20 replied:
Will add logic to check if the channel is closed or not

Aneurysm9 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go:
```diff
@@ -0,0 +1,134 @@
+
+type taskInfoOption func(*taskInfo)
+
+func newECSTaskInfo(ctx context.Context, ecsTaskEndpointProvider hostIPProvider,
```
All of the same comments from the instance info file seem to apply here as well.

JohnWu20 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,136 @@
+const (
+	defaultMaxRetries              = 3
+	defaultTimeout                 = 1 * time.Second
+	defaultBackoffRetryBaseInMills = 200
```
Thanks will change it

JohnWu20 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,136 @@
+
+type HTTPClient struct {
+	maxRetries              int
+	backoffRetryBaseInMills time.Duration
```
Thanks will change it

JohnWu20 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (h *HTTPClient) backoffSleep(currentRetryCount int) {
+	backoffInMillis := int64(float64(h.backoffRetryBaseInMills) * math.Pow(2, float64(currentRetryCount)))
+	sleepDuration := time.Millisecond * time.Duration(backoffInMillis)
```
Thanks, will change it.

JohnWu20 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,136 @@
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(ctx, endpoint)
+		if err != nil {
+			sugar := logger.Sugar()
+			sugar.Warn("W! retry [%d/%d], unable to get http response from %s, error: %v", i, h.maxRetries, endpoint, err)
```
Thanks, will change it

JohnWu20 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,128 @@
+
+	shouldRefresh := func() bool {
+		//stop the refresh once we get instance ID and type successfully
+		return cii.clusterName == "" || cii.containerInstanceID == ""
```
Thanks, will change it

JohnWu20 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+func (cii *containerInstanceInfo) GetContainerInstanceID() string {
+	cii.Lock()
+	defer cii.Unlock()
```
Will change it, thanks

JohnWu20 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+func (cii *containerInstanceInfo) GetClusterName() string {
+	cii.Lock()
+	defer cii.Unlock()
```
Will change it, thanks

JohnWu20 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,128 @@
+
+type ecsInstanceInfoOption func(*containerInstanceInfo)
+
+func newECSInstanceInfo(ctx context.Context, ecsAgentEndpointProvider hostIPProvider,
+	refreshInterval time.Duration, logger *zap.Logger, httpClient Requester, readyC chan bool, options ...ecsInstanceInfoOption) containerInstanceInfoProvider {
```
Will change it, thanks

JohnWu20 commented on internal/aws/containerinsight/httpclient/httpclient_test.go:
```diff
@@ -0,0 +1,89 @@
+	return f.response, nil
+}
+
+func TestRequestSecuess(t *testing.T) {
```
Will change it, thanks

JohnWu20 committed:
Using retriablehttp and revising the code

JohnWu20 committed:
Using retriablehttp and revising the code

JohnWu20 committed:
Merge branch 'ecsinstancemetrics' of https://github.com/JohnWu20/opentelemetry-collector-contrib into ecsinstancemetrics

JohnWu20 committed:
rebase

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (h *HTTPClient) clientGet(ctx context.Context, url string) (resp *http.Response, err error) {
+	req, err := retryablehttp.NewRequest("GET", url, nil)
```
```suggestion
	req, err := retryablehttp.NewRequest(http.MethodGet, url, nil)
```

Aneurysm9 commented on internal/aws/containerinsight/httpclient/httpclient_test.go:
```diff
@@ -0,0 +1,144 @@
+
+	respBody := "body"
+	response := &http.Response{
+		Status:        "200 OK",
```
```suggestion
		Status:        "Bad Request",
```

Aneurysm9 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:
```diff
@@ -0,0 +1,124 @@
+	err = json.Unmarshal(resp, containerInstance)
+	if err != nil {
+		cii.logger.Warn("Failed to unmarshal: ", zap.Error(err))
+		cii.logger.Warn("Resp content is " + string(resp))
```
```suggestion
		cii.logger.Debug("Resp content is " + string(resp))
```

This may be a bit much to log, up to 5MB.

Aneurysm9 commented on receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go:
```diff
@@ -0,0 +1,132 @@
+	err = json.Unmarshal(resp, ecsTasksInfo)
+	if err != nil {
+		ti.logger.Warn("Unable to parse resp from ecsagent taskinfo endpoint, error:", zap.Error(err))
+		ti.logger.Warn("D! resp content is %s" + string(resp))
```
```suggestion
		ti.logger.Debug("D! resp content is %s" + string(resp))
```

JohnWu20 committed:
Update receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

JohnWu20 committed:
Update receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

JohnWu20 committed:
Update internal/aws/containerinsight/httpclient/httpclient_test.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

JohnWu20 committed:
Update internal/aws/containerinsight/httpclient/httpclient.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

JohnWu20 committed:
Update internal/aws/containerinsight/httpclient/httpclient.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

bogdandrutu commented on internal/aws/containerinsight/httpclient/httpclient.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+type doer interface {
+	Do(request *retryablehttp.Request) (*http.Response, error)
```
We do have already a queued-retry logic for exporters, why is this needed?

JohnWu20 replied:
This is used for calling the ECS container agent introspection task endpoint and cluster endpoint, and there are some specific setting for the response data, I think this is different from the logic for exporters.

bogdandrutu replied:
@Aneurysm9 you know both, please followup if this is a duplicate logic to ensure we use the exporter helper if possible :)
